smolyay/samples.py

```diff
@@ -508,3 +511,583 @@ def _create(self):
+    ----------
+    unique_points : list of :class:UnidimensionalPointSet
+        unique set of 1D points
+    """
+
+    def _create(self):
+        """Generating the tensor grid
+
+        Using the parameters of the class, generate a grid set using all
+        combinations of unidimensional points
+        """
+        num_points = numpy.prod([len(p) for p in self._point_sets])
+        self._points = numpy.zeros((num_points, len(self._point_sets)), dtype=float)
+        for i, point in enumerate(itertools.product(*self._point_sets)):
+            self._points[i] = point
+
+class SmolyakSparseProductPointSet(PointSetProduct):
+    """Generate a grid using sparse combinations of unidimensional point sets
+
+    Depending on the dimensionality, points and provided by
+    :class:`UnidimensionalPointSet`, make full tensor grids.
+    :meth:`generates_points` generates a full tensor grid
+
+    Parameters
+    ----------
+    unique_points : list of :class:UnidimensionalPointSet
+        unique set of 1D points
+    """
+
+    def _create(self):
+        """Generating the tensor grid
+
+        Using the parameters of the class, generate a grid set using all
+        combinations of unidimensional points
+        """
+        # reset points
+        self._points = None
+        # find number of levels per dimension
+        num_levels_per_dim = [ob.num_levels for ob in self._point_sets]
+        max_num_levels = max(num_levels_per_dim)
+        # get the combinations of levels based on maximum possible number of levels
+        level_combinations = []
+        for sum_of_levels in range(max_num_levels):
+            level_combinations.extend(
+                list(
+                    generate_compositions(
+                        sum_of_levels, self.num_dimensions, include_zero=True
+                    )
+                )
+            )
+        level_combinations = numpy.array(level_combinations)
+        # remove combinations where a dimension exceeds its number of levels
+        # only check if point sets have different numbers of levels
+        if min(num_levels_per_dim) != max_num_levels:
+            valid_comb = numpy.all(
+                numpy.greater(num_levels_per_dim, level_combinations), axis=1
+            )
+            level_combinations = level_combinations[valid_comb, :]
+        # generate sets of points based on combinations of levels
+        for level_comb in level_combinations:
+            level_point_combinations = [
+                self._point_sets[d].level(level) for d, level in enumerate(level_comb)
+            ]
+            points_ = numpy.array(numpy.meshgrid(*level_point_combinations)).T.reshape(
+                -1, self.num_dimensions
+            )
+            if self._points is None:
+                self._points = points_
+            else:
+                self._points = numpy.concatenate((self._points, points_), axis=0)
+
+
+def generate_compositions(value, num_parts, include_zero):
```

Comment:
Make this private with a leading underscore, I don't think we want to guarantee it as part of the API.
```suggestion
def _generate_compositions(value, num_parts, include_zero):
```